cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TypeManager.kt

```diff
@@ -187,13 +190,35 @@ class TypeManager {
         return parameterizedType
     }
 
-    fun <T : Type> registerType(t: T): T {
+    inline fun <reified T : Type> registerType(t: T): T {
+        // Skip as they should be unique to each class and not globally unique
+        if (t is ParameterizedType) {
+            return t
+        }
+
         if (t.isFirstOrderType) {
-            firstOrderTypes.add(t)
-        } else {
-            secondOrderTypes.add(t)
-            registerType((t as SecondOrderType).elementType)
+            // Make sure we only ever return one unique object per type
+            if (!firstOrderTypes.add(t)) {
+                return firstOrderTypes.filterIsInstance<T>().first { it == t }
```

Comment:
Doesn't filterIsInstance<T> return the same type of the same type type but not the one that was found in the condition above that is equal with `.equals`?